### ISSUE TITLE
Rule proposal: `better-boolean-var-declaration`

### ISSUE THREAD

**Issue by axetroy** [sindresorhus/eslint-plugin-unicorn#2441]
### Description

There are many irregular variable declarations in some Boolean variable declarations.

For example

```js
const showModal = false
```

This value is obviously a Boolean value, but it cannot be seen from the variable.

Standard variable declaration can improve code readability

Boolean variables should start with 'is', 'was', 'has', 'can', 'should', etc.

### Fail

```js
const showModal = false

const greatThen = a > b

const completed = progress === 100

const setProperty = key in object

const stringType = typeof input === 'string'

// ...
```

### Pass

```js
const isShowModal = false

const isGreatThen = a > b

const isCompleted = progress === 100

const hasSetProperty = key in object

const isStringType = typeof input === 'string'
```

### Proposed rule name

better-boolean-var-declaration

### Additional Info

This rule has been implemented in my company project, it can't autofix, but it can provide suggestions. I think it is useful

If this is acceptable, I'd be happy to PR

<details>
<summary>See the test case</summary>

```js
const assert = require('node:assert')
const test = require('node:test')

const outdent = require('outdent')
const { RuleTester } = require('eslint')

const rule = require('./boolean-var')

const ruleTester = new RuleTester({
    languageOptions: {
        ecmaVersion: 'latest',
        parserOptions: {
            sourceType: 'module'
        }
    }
})

test('boolean-var', t => {
    assert.strictEqual(typeof rule.create, 'function')

    ruleTester.run('boolean-var', rule, {
        valid: [
            // Valid cases
            {
                code: 'var isCompleted = true'
            },
            {
                code: 'var _isCompleted = true'
            },
            {
                code: 'var __isCompleted = true'
            },
            {
                code: 'var isCompleted = Boolean(0)'
            },
            {
                code: 'var isCompleted = !0'
            },
            {
                code: 'var isCompleted = !!0'
            },
            {
                code: 'var willUpdate = true; var allowsUpdate = true',
                options: [{ prefixes: ['will', 'allows'] }]
            }
        ],

        invalid: [
            // Invalid cases
            {
                code: 'var completed = true',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: `var ${prefix}Completed = true`
                            }
                        })
                    }
                ]
            },
            {
                code: 'var completed = "true"',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: `var ${prefix}Completed = "true"`
                            }
                        })
                    }
                ]
            },
            {
                code: 'var SKIP_REMIND = true',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'SKIP_REMIND',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix.toUpperCase()}_SKIP_REMIND\``,
                                output: `var ${prefix.toUpperCase()}_SKIP_REMIND = true`
                            }
                        })
                    }
                ]
            },
            {
                code: outdent`
                  var completed = true

                  console.log(completed)

                  function foo (completed) {
                    console.log(completed)
                  }

                  function bar () {
                      console.log(completed)
                  }
                `,
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: outdent`
                                  var ${prefix}Completed = true
              
                                  console.log(${prefix}Completed)
              
                                  function foo (completed) {
                                    console.log(completed)
                                  }
              
                                  function bar () {
                                      console.log(${prefix}Completed)
                                  }
                                `
                            }
                        })
                    }
                ]
            },
            {
                code: 'var completed = Boolean(0)',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: `var ${prefix}Completed = Boolean(0)`
                            }
                        })
                    }
                ]
            },
            {
                code: 'var completed = !0',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: `var ${prefix}Completed = !0`
                            }
                        })
                    }
                ]
            },
            {
                code: 'var completed = !!0',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'completed',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Completed\``,
                                output: `var ${prefix}Completed = !!0`
                            }
                        })
                    }
                ]
            },
            {
                code: 'class Man { detected = false }',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'detected',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Detected\``,
                                output: `class Man { ${prefix}Detected = false }`
                            }
                        })
                    }
                ]
            },
            {
                code: 'class Man { static detected = false }',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'detected',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Detected\``,
                                output: `class Man { static ${prefix}Detected = false }`
                            }
                        })
                    }
                ]
            },
            {
                code: 'class Man { #detected = false }',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'detected',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Detected\``,
                                output: `class Man { #${prefix}Detected = false }`
                            }
                        })
                    }
                ]
            },
            {
                code: 'const a = b && c',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'a',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}A\``,
                                output: `const ${prefix}A = b && c`
                            }
                        })
                    }
                ]
            },
            {
                code: 'const func = typeof fn === "function"',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'func',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Func\``,
                                output: `const ${prefix}Func = typeof fn === "function"`
                            }
                        })
                    }
                ]
            },
            {
                code: 'const defined = key in object',
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'defined',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Defined\``,
                                output: `const ${prefix}Defined = key in object`
                            }
                        })
                    }
                ]
            },
            {
                code: outdent`
                  const defined = key in object

                  const obj = {
                    defined
                  }
                `,
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'defined',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Defined\``,
                                output: outdent`
                                  const ${prefix}Defined = key in object
              
                                  const obj = {
                                    defined: ${prefix}Defined
                                  }
                                `
                            }
                        })
                    }
                ]
            },
            {
                code: outdent`
                  const defined = key in object

                  export default { defined }
                `,
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'defined',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Defined\``,
                                output: outdent`
                                  const ${prefix}Defined = key in object
              
                                  export default { defined: ${prefix}Defined }
                                `
                            }
                        })
                    }
                ]
            },
            {
                code: outdent`
                  const isFoo = true

                  const defined = isFoo

                  export default { defined }
                `,
                errors: [
                    {
                        messageId: 'missingPrefix',
                        data: {
                            name: 'defined',
                            prefixes: '`is`, `was`, `has`, `can`, or `should`'
                        },
                        suggestions: ['is', 'was', 'has', 'can', 'should'].map(prefix => {
                            return {
                                desc: `Rename to \`${prefix}Defined\``,
                                output: outdent`
                                  const isFoo = true

                                  const ${prefix}Defined = isFoo
              
                                  export default { defined: ${prefix}Defined }
                                `
                            }
                        })
                    }
                ]
            }
        ]
    })
})

```
</details>

**Comment by fregante** [comment]
Duplicate of https://github.com/sindresorhus/eslint-plugin-unicorn/issues/617

**Comment by sindresorhus** [comment]
Just for later docs, this is not a good example:

```js
const isShowModal = false
```

Should be:

```js
const isShowingModal = false
```

or

```js
const isModalPresented = false
```

**Comment by axetroy** [comment]
> Overlap with:
> 
>     * [Only functions names can have verbs (get, set, create, remove) #617](https://github.com/sindresorhus/eslint-plugin-unicorn/issues/617)
> 
>     * [Rule proposal: Ensure special verbs match the type (e.g. `hasName` is boolean or boolean function) typescript-eslint/typescript-eslint#3032](https://github.com/typescript-eslint/typescript-eslint/issues/3032)

I think their functions are similar, but not overlapping.

Maybe they can be combined into one rule `better-var-naming`?

**Comment by fregante** [comment]
The entire logic is overlapping: identifier must follow type. That one mainly targeted functions while this one is for booleans.

**Comment by axetroy** [comment]
@fregante  I plan to implement this.

But I'm not sure if I should merge these into one rule.

Because these look similar, But writing it into one rule will increase the complexity a lot, which may be detrimental to maintenance.

What do you think?

**Comment by fregante** [comment]
Yeah even if they share some/most logic, it might be easier to implement them separately.

**Comment by axetroy** [comment]
Before I start, I need to make sure that the property also needs to be checked, as it cannot be safely repaired.

In my current implementation, I have intentionally omitted it because it is not safe

```js
const downloadTask = {
	completed: progress === 100
}
```

**Comment by fregante** [comment]
In my view this rule is too restrictive. Like I had suggested in https://github.com/sindresorhus/eslint-plugin-unicorn/issues/617, it should _block_ incorrect usage, not _require_ certain verbs. Much easier to report `wasRed = 2` as being the wrong type than understand that `requiresLogin = true` is totally perfectly fine as a boolean name.

TL;DR: block booleanish verbs on non-boolean types instead.
